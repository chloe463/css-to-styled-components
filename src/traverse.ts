import { ChildNode } from "postcss";
import { capitalize } from "./utils/cases";

export const traverse = (node: ChildNode, indentLevel = 0) => {
  switch (node.type) {
    case "atrule": {
      return;
    }
    case "decl": {
      if (node.prop.match(/^\$[a-zA-Z0-9]+/)) {
        return `const ${node.prop.replace("$", "")} = "${node.value}";`;
      }
      return `${node.prop}: ${node.value};`;
    }
    case "rule": {
      const declarations: (string | undefined)[] = node.nodes.map(v => traverse(v, indentLevel+1));
      let [element, name] = node.selector.split(/\.|#/);
      const styledFnType = `styled.${element || "div"}`;
      if (indentLevel) {
        const indent = Array.from({ length: indentLevel * 2 }, _ => " ").join("");
        return [
          `${node.selector} {`,
          declarations.map(d => `  ${indent}${d}`).join("\n"),
          `${indent}}`,
        ].join("\n");
      }
      return [
        `const ${capitalize(name)} = ${styledFnType}\``,
        declarations.map(d => `  ${d}`).join("\n"),
        `\`;`,
      ].join("\n");
    }
    case "comment": {
      return;
    }
  }
};
