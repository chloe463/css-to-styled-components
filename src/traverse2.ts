// This function is forked from https://github.com/salesforce-ux/scss-parser/blob/master/lib/stringify.js

import _ from "lodash";
import type { Node } from "scss-parser";
import { kebabToCamel } from "./utils/cases";

type toStringFn = (n: Node) => string;
interface TypeMappings {
  [key: string]: toStringFn;
}

const type: TypeMappings = {
  arguments: (n) =>
    '(' + walkValue(n.value) + ')',
  atkeyword: (n) =>
    '@' + n.value,
  attribute: (n) =>
    '[' + walkValue(n.value) + ']',
  block: (n) =>
    '{' + walkValue(n.value) + '}',
  class: (n) =>
    'export const ' + walkValue(n.value),
  color_hex: (n) =>
    '#' + n.value,
  id: (n) =>
    '#' + walkValue(n.value),
  interpolation: (n) =>
    '#{' + walkValue(n.value) + '}',
  comment_multiline: (n) =>
    '/*' + n.value + '*/',
  comment_singleline: (n) =>
    '//' + n.value,
  parentheses: (n) =>
    '(' + walkValue(n.value) + ')',
  pseudo_class: (n) =>
    ':' + walkValue(n.value),
  psuedo_element: (n) =>
    '::' + walkValue(n.value),
  selector: (n) => {
    walkSelector(n);
    return "";
    // return 'export const ' + walkSelector(n);
  },
  string_double: (n) =>
   `"${n.value}"`,
  string_single: (n) =>
   `'${n.value}'`,
  variable: (n) =>
    '$' + n.value
}

export const walkNode = (node: Node): string => {
  if (type[node.type]) return type[node.type](node)
  if (_.isString(node.value)) return node.value
  if (_.isArray(node.value)) return walkValue(node.value)
  return ''
}

export const walkSelector = (node: Node) => {
  if (_.isArray(node.value)) {
    const ids = node.value
     .filter((v) => v.type === "id")
     .map((v) => _.isArray(v.value) ? v.value[0].value : v.value)
     .join("");
    const classes = node.value.filter((v) => v.type === "class");
    const varName = classes.map(v => _.isArray(v.value) ? kebabToCamel(v.value[0].value as string) : kebabToCamel(v.value)).join("");
    const identifiers = node.value.filter((v) => v.type === "identifier");
    console.log({ varName, ids, identifiers });
    if (ids.length > 0) {
    }
    return "";
  }
};

export const walkValue = (value: string | Node[]) => {
  if (!_.isArray(value)) return ''
  return value.reduce((s, node) => {
    return s + walkNode(node)
  }, '')
}
