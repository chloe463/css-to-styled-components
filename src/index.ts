import fs from "fs";
import postcss from "postcss";
import { traverse } from './traverse';

const filePath = process.argv[2];

if (!filePath) {
  console.error(`Please give a css file name.`);
  process.exit(1);
}

if (!fs.existsSync(filePath)) {
  console.error(`${filePath} does not exists`);
  process.exit(1);
}

const scss = fs.readFileSync(filePath, "utf8").replace(/\s*\/\/.*/g, "");

const ast = postcss.parse(scss);

if (!ast || !ast.source) {
  console.error(`Failed to parse file content. The given file is invalid css file.`);
  process.exit(1);
}

const result = ast.nodes.map((node) => traverse(node, 0));
console.log(result.join("\n"));

// const sources = ast.source.input.css.split("\n");

// const result = ast.nodes.map(node => {
//   if (node.type === "atrule" && node.name === "import") {
//     return;
//   }

//   let name = "";
//   let styledFnType = `styled.div\``;
//   let defs: string[] = [];
//   if (node.type === "atrule" && node.name === "keyframes") {
//     styledFnType = `keyframes\``;
//   } else if (node.type === "rule") {
//     name = node.selector
//       .replace(/^\./g, "")
//       .replace(/-([a-z])/g, (_, p) => p.toUpperCase());

//     const start = node.source?.start?.line;
//     const end = node.source?.end?.line;
//     defs = start && end ? sources.slice(start, end - 1) : [];
//   } else if (node.type === "decl") {
//     return ["const", node.prop.replace(/^\$/, ""), "=", `"${node.value}";`].join(" ");
//   }

//   return [
//     `const ${capitalize(name)} = ${styledFnType}`,
//     ...defs,
//     `\``,
//   ].join("\n");
// });

// console.log(
//   result
//     .join("\n")
//     .replace(/:.*\s+\$(.*)\;/g, (_, p) => {
//       const cameled = p.replace(/-([a-z])/g, (_: number, p: string) => p.toUpperCase());
//       return `: \${${cameled}};`;
//     })
//     .replace(/composes:\s*([a-zA-Z0-9-]+)(\sfrom.*)?/g, (match, p) => {
//       return [
//         `// ${match}`,
//         "  ${" + snakeToCamel(kebabToCamel(p)) + "}",
//       ].join("\n");
//     })
// );
