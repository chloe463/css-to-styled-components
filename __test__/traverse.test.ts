import { AtRule, Declaration, Rule } from "postcss";
import { traverse } from "../src/traverse";

describe("traverse", () => {
  describe("can transform css decl to styled-components", () => {
    it ("transform variable decl", () => {
      const node = new Declaration({ prop: "$color", value: "black" });
      expect(traverse(node)).toMatchSnapshot();
    });
  });
  describe("can transform css rule to styled-components", () => {
    it ("with no element", () => {
      const node = new Rule({
        selector: ".Klass",
        nodes: [
          new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.86)" }),
          new Declaration({ prop: "color", value: "rgb(255, 255, 255)" }),
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    });

    it ("with element", () => {
      const node = new Rule({
        selector: "p.message",
        nodes: [
          new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.86)" }),
          new Declaration({ prop: "color", value: "rgb(255, 255, 255)" }),
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    });

    it ("nested rules", () => {
      const node = new Rule({
        selector: "div.container",
        nodes: [
          new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.86)" }),
          new Declaration({ prop: "color", value: "rgb(255, 255, 255)" }),
          new Rule({
            selector: "&:hover",
            nodes: [
              new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.74)" }),
              new Declaration({ prop: "color", value: "$white-alpha500" }),
            ],
          }),
          new Rule({
            selector: "span.message",
            nodes: [
              new Declaration({ prop: "font-weight", value: "400" }),
              new Rule({
                selector: "p",
                nodes: [
                  new Declaration({ prop: "color", value: "rgba(0, 0, 0, 0.56)" }),
                  new Rule({
                    selector: "b",
                    nodes: [
                      new Declaration({ prop: "font-weight", value: "600" }),
                      new Declaration({ prop: "color", value: "rgba(0, 0, 0, 0.74)" }),
                    ],
                  }),
                  new Rule({
                    selector: "i",
                    nodes: [
                      new Declaration({ prop: "font-weight", value: "400" }),
                      new Declaration({ prop: "color", value: "rgba(0, 0, 0, 0.74)" }),
                      new Declaration({ prop: "text-decoration", value: "underline" }),
                    ],
                  })
                ],
              }),
            ],
          }),
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    });

    it ("keyframes", () => {
      const node = new AtRule({
        name: "keyframes",
        params: "anim",
        nodes: [
          new Rule({
            selector: "0%",
            nodes: [
              new Declaration({ prop: "opacity", value: "0" }),
            ],
          }),
          new Rule({
            selector: "100%",
            nodes: [
              new Declaration({ prop: "opacity", value: "1" }),
            ],
          })
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    })
  });
});
