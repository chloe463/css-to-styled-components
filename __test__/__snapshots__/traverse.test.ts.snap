// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverse can transform css decl to styled-components transform variable decl 1`] = `"const color = \\"black\\";"`;

exports[`traverse can transform css rule to styled-components nested rules 1`] = `
"const Container = styled.div\`
  background-color: rgba(0, 0, 0, 0.86);
  color: rgb(255, 255, 255);
  &:hover {
    background-color: rgba(0, 0, 0, 0.74);
    color: rgb(250, 250, 250);
  }
  span.message {
    font-weight: 400;
    p {
      color: rgba(0, 0, 0, 0.56);
      b {
        font-weight: 600;
        color: rgba(0, 0, 0, 0.74);
      }
      i {
        font-weight: 400;
        color: rgba(0, 0, 0, 0.74);
        text-decoration: underline;
      }
    }
  }
\`;"
`;

exports[`traverse can transform css rule to styled-components with element 1`] = `
"const Message = styled.p\`
  background-color: rgba(0, 0, 0, 0.86);
  color: rgb(255, 255, 255);
\`;"
`;

exports[`traverse can transform css rule to styled-components with no element 1`] = `
"const Klass = styled.div\`
  background-color: rgba(0, 0, 0, 0.86);
  color: rgb(255, 255, 255);
\`;"
`;
